File: karatsuba.py
def karatsuba(a,b):
    """\nfunction to multiply two number's using karatsuba algorithm in a more efficient TC=n^1.585 \n"""
    if(len(str(a))==1 or len(str(b))==1):
        return a*b
    else:
        d=max(len(str(a)),len(str(b)))
        newd=d//2

        # a=x1.10^(newd)+x0
        # b=y1.10^(newd)+y0

        x1=a//(10**newd)
        x0=a%(10**newd)
        y1=b//(10**newd)
        y0=b%(10**newd)

        x0y0=karatsuba(x0,y0)
        x1y1=karatsuba(x1,y1)
        xiyj=karatsuba(x1+x0,y1+y0)-x0y0-x1y1
        # T(n)=3T(n/2)+Θ(n)

        return x1y1*(10**(2*newd))+x0y0+xiyj*(10**newd)

# print(karatsuba.__doc__) to print docstring
print(karatsuba(5484654534,2185146235))

File: karatsuba.cpp
//Multiplying two big numbers using karatsuba algorithm(divide and conquer approach)
#include<bits/stdc++.h>
#include <cstdio>
#include <sstream>
using namespace std;
#define for(i,a,b) for(int i=0;i<n;i++)
typedef long long int ll; 

ll multi(ll x,ll y){
    stringstream ss,ss1;
    string s;
    ss<<x;
    ss>>s;
    int l1=s.length();
    ss1<<y;
    ss1>>s;
    int l2=s.length();
    
    int d=max(l1,l2)/2;
    ll x1=x/pow(10,d);
    ll x0=x % (ll)pow(10,d);
    ll y1=y / pow(10,d);
    ll y0=y % (ll)pow(10,d);

    ll x0y0=x0*y0;
    ll x1y1=x1*y1;
    ll x1y0_x0y1=(x0+x1)*(y0+y1)-x0y0-x1y1;
    return x1y1*pow(10,2*d)+x1y0_x0y1*pow(10,d)+x0y0;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    // std::ifstream input{"input.txt"};
    FILE *f1,*f2;
    f1=fopen("input.txt","r");
    f2=fopen("output.txt","w");
    ll x,y;
    fscanf(f1,"%lld %lld",&x,&y);
    fprintf(f2,"%lld",multi(x, y));
    return 0;
}